pyat/at/latticetools/observables.py

```diff
@@ -0,0 +1,974 @@
+
+
+class _ArrayAccess(object):
+    """Access to selected items in an array"""
+
+    def __init__(self, index):
+        self.index = _all_rows(index)
+
+    def __call__(self, ring, data):
+        index = self.index
+        return data if index is None else data[self.index]
+
+
+class _RecordAccess(object):
+    """Access to selected items in a record array"""
+
+    def __init__(self, fieldname: str, index):
+        self.index = index
+        self.fieldname = fieldname
+
+    def __call__(self, ring, data):
+        index = self.index
+        data = getattr(data, self.fieldname)
+        return data if index is None else data[self.index]
+
+
+def _all_rows(index: Optional[RefIndex]):
+    """Prepends "all rows" (":") to an index tuple"""
+    if index is None:
+        return None
+    if isinstance(index, tuple):
+        return (slice(None),) + index
+    else:
+        return slice(None), index
+
+
+class _Tune(object):
+    """Get integer tune from the phase advance"""
+
+    def __init__(self, idx: RefIndex):
+        self.fun = _RecordAccess("mu", _all_rows(idx))
+
+    def __call__(self, ring, data):
+        mu = self.fun(ring, data)
+        return np.squeeze(mu, axis=0) / 2 / pi
+
+
+class _Ring(object):
+    """Get an attribute of a lattice element"""
+
+    def __init__(self, attrname, index, refpts):
+        self.get_val = _RecordAccess(attrname, index)
+        self.refpts = refpts
+
+    def __call__(self, ring):
+        vals = [self.get_val(ring, el) for el in ring.select(self.refpts)]
+        return np.array(vals)
+
+
+class Need(Enum):
+    """Defines the computation requirements for an :py:class:`Observable`."""
+
+    #:  Specify :py:func:`.find_orbit` computation and provide its *orbit*
+    #:  t to the evaluation function
+    ORBIT = 1
+    #:  Specify :py:func:`.find_m44` or :py:func:`.find_m44` computation and
+    #:  provide its *m44* or *m66* output to the evaluation function
+    MATRIX = 2
+    #:  Specify :py:func:`.get_optics` computation and provide its *ringdata*
+    #:  to the evaluation function
+    GLOBALOPTICS = 3
+    #:  Specify :py:func:`.get_optics` computation and provide its *elemdata*
+    #:  to the evaluation function
+    LOCALOPTICS = 4
+    #:  Specify :py:func:`.lattice_pass` computation and provide its *r_out*
+    #:  to the evaluation function
+    TRAJECTORY = 5
+    #:  Specify :py:func:`.envelope_parameters` computation and provide its
+    #:  *params* to the evaluation function
+    EMITTANCE = 6
+    #:  Associated with LOCALOPTICS, require local optics computation at all
+    #:  points: slower but avoids jumps in phase advance
+    ALL_POINTS = 7
+    #:  Associated with LOCALOPTICS, require the *get_chrom* keyword
+    CHROMATICITY = 8
+    #:  Associatef with LOCALOPTICS, require the *get_w* keyword
+    W_FUNCTIONS = 9
+    #:  Specify geometry computation and provide the full data at evaluation points
+    GEOMETRY = 10
+
+
+class Observable(object):
```

Comment:
Could be done… But since this can be used directly by users, I find the simple `Observable` name clearer. I can change if you find ObservableBase better…